src/lib/frontend/d_cnf.ml

```diff
@@ -924,11 +923,144 @@ let rec mk_expr ?(loc = Loc.dummy) ?(name_base = "")
+
+          | B.Bitv_neg n, [t] ->
+            E.mk_term (Sy.Op BVneg) [aux_mk_expr t] (Ty.Tbitv n)
+
+          | B.Bitv_not n, [t] ->
+            E.mk_term (Sy.Op BVnot) [aux_mk_expr t] (Ty.Tbitv n)
+
           (* Binary applications *)
 
           | B.Bitv_concat { n; m; }, [ x; y ] ->
-            let rty = Ty.Tbitv (n+m) in
-            E.mk_term (Sy.Op Sy.Concat) [aux_mk_expr x; aux_mk_expr y] rty
+            E.mk_bitv_concat (aux_mk_expr x) (aux_mk_expr y) (n + m)
+
+          | B.Bitv_or n, [ x; y ] ->
+            E.mk_term
+              (Sy.Op Sy.BVor) [aux_mk_expr x; aux_mk_expr y] (Ty.Tbitv n)
+
+          | B.Bitv_and n, [ x; y ] ->
+            E.mk_term
+              (Sy.Op Sy.BVand) [aux_mk_expr x; aux_mk_expr y] (Ty.Tbitv n)
+
+          | B.Bitv_nor n, [ x; y ] ->
+            E.mk_term
+              (Sy.Op Sy.BVnor) [aux_mk_expr x; aux_mk_expr y] (Ty.Tbitv n)
+
+          | B.Bitv_nand n, [ x; y ] ->
+            E.mk_term
+              (Sy.Op Sy.BVnand) [aux_mk_expr x; aux_mk_expr y] (Ty.Tbitv n)
+
+          | B.Bitv_xor n, [ x; y ] ->
+            E.mk_term
+              (Sy.Op Sy.BVxor) [aux_mk_expr x; aux_mk_expr y] (Ty.Tbitv n)
+
+          | B.Bitv_xnor n, [ x; y ] ->
+            E.mk_term
+              (Sy.Op Sy.BVxnor) [aux_mk_expr x; aux_mk_expr y] (Ty.Tbitv n)
+
+          | B.Bitv_comp n, [ x; y ] ->
+            E.mk_term
+              (Sy.Op Sy.BVcomp) [aux_mk_expr x; aux_mk_expr y] (Ty.Tbitv n)
+
+          | B.Bitv_add n, [ x; y ] ->
+            E.mk_term
+              (Sy.Op Sy.BVadd) [aux_mk_expr x; aux_mk_expr y] (Ty.Tbitv n)
+
+          | B.Bitv_sub n, [ x; y ] ->
+            E.mk_term
+              (Sy.Op Sy.BVsub) [aux_mk_expr x; aux_mk_expr y] (Ty.Tbitv n)
+
+          | B.Bitv_mul n, [ x; y ] ->
+            E.mk_term
+              (Sy.Op Sy.BVmul) [aux_mk_expr x; aux_mk_expr y] (Ty.Tbitv n)
+
+          | B.Bitv_udiv n, [ x; y ] ->
+            E.mk_term
+              (Sy.Op Sy.BVudiv) [aux_mk_expr x; aux_mk_expr y] (Ty.Tbitv n)
+
+          | B.Bitv_urem n, [ x; y ] ->
+            E.mk_term
+              (Sy.Op Sy.BVurem) [aux_mk_expr x; aux_mk_expr y] (Ty.Tbitv n)
+
+          | B.Bitv_sdiv n, [ x; y ] ->
+            E.mk_term
+              (Sy.Op Sy.BVsdiv) [aux_mk_expr x; aux_mk_expr y] (Ty.Tbitv n)
+
+          | B.Bitv_srem n, [ x; y ] ->
+            E.mk_term
+              (Sy.Op Sy.BVsrem) [aux_mk_expr x; aux_mk_expr y] (Ty.Tbitv n)
+
+          | B.Bitv_smod n, [ x; y ] ->
+            E.mk_term
+              (Sy.Op Sy.BVsmod) [aux_mk_expr x; aux_mk_expr y] (Ty.Tbitv n)
+
+          | B.Bitv_shl n, [ x; y ] ->
+            E.mk_term
+              (Sy.Op Sy.BVshl) [aux_mk_expr x; aux_mk_expr y] (Ty.Tbitv n)
+
+          | B.Bitv_lshr n, [ x; y ] ->
+            E.mk_bvlshr n (aux_mk_expr x) (aux_mk_expr y)
+
+          (* BV operators that produce ITEs or booleans are partially expanded
+             here instead of in the theory. *)
```

Comment:
Shostak seemed to not like receving ITEs as the output of theorie's `make` function.
But I have to check, there might be something I missed or it might be due to the split of the make function into two `make` and `make_aux`/`make_rec`, one of which produces `r` terms and thte other produces a list of `r simple_term alpha_term`. But yeah, needs a bit more investigation.